Add unit tests to increase coverage

lennysgarage opened the pull request:
Fixing failed magic mirror PR: https://github.com/open-cluster-management-io/multicloud-operators-subscription/pull/240


* Add tests for helmrelease_types

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add tests for ansiblejob_types

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Remove unused type deployable

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Remove deployable references

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test case for downloadGitRepo

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test for GetKnownHostFromURL

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test for GetCertChain

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test case for DownloadChartFromHelmRepoHTTP

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test case for DownloadGitRepo

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Remove unused func CompareOjbectStoreStatus

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add tests for SubscribeItem & UnsubscribeItem

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add tests for subscription with a bucket

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add tests for SetSuffix & SetInClusterPackageStatus

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add tests for IsHub & IsSubscriptionBeDeleted

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add tests for AllowApplyTemplate & OverrideResourceBySubscription

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Didn't meant to commit cover.html

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Fix IsSubscriptionBeDeleted test

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test for IfClusterAdmin

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Remove unused funcs Base64StringDecode & GetManagedClusterNamespace

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Remove unused func Map

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Remove failing test IsSubscriptionBeDeleted

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test for PlacementDecisionPredicateFunc

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test for ManagedClusterSecretPredicateFunc

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

* Add test case for ArgocdServerPredicateFunc

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

mikeshng commented:
@lennysgarage could you double check why magic mirror was not able to merge this change? 

Did we go out of sync from upstream?

lennysgarage commented:
> 
Hey Mike, so I've looked further into the problem here and it's not that we've gone out of sync, but that magic-mirror is cherry-picking our commits in a way that doesn't work for us.

In the case here, magic-mirror tried to cherry-pick 24 commits from the merged upstream PR using this command:
`git cherry-pick -x fd29e047a161c0ff5f39fa9b14c2f9db70b4c0f0~24..fd29e047a161c0ff5f39fa9b14c2f9db70b4c0f0 --allow-empty --keep-redundant-commits`

However the command is not actually taking the 24 commits from inside of the upstream PR, but instead cherry-picking the 24 commits before the merged PR. This is where the problem lies, because there's so many commits its cherry-picking it goes back and hits this problem commit, https://github.com/open-cluster-management-io/multicloud-operators-subscription/commit/82ac6c6dddbb17b83ada4d29d931ef8bde43b944. This commit contains files that the midstream does not contain that we removed to fit, namely the ` .github/workflows` contents. So when magic-mirror tries to cherry-pick this commit it encounters a conflict and is unable to automatically create the PR.
`CONFLICT (modify/delete): .github/workflows/go-postsubmit.yml deleted in HEAD and modified in f6413df (update go to 1.18).  Version f6413df (update go to 1.18) of .github/workflows/go-postsubmit.yml left in tree.`

I've found this also follows previous failed and non failed PRs. The PR right before this [one](https://github.com/open-cluster-management-io/multicloud-operators-subscription/pull/241) only has a single commit and so doesn't reach the problem. While an older PR like: https://github.com/open-cluster-management-io/multicloud-operators-subscription/pull/228 contained 13 commits and so while being cherry-picked hit the problem at commit 10.
You can test this by running 
`git cherry-pick -x 54fbd317f22ab072fe68488ff2375720465a844a~9..54fbd317f22ab072fe68488ff2375720465a844a --allow-empty --keep-redundant-commits` compared to `git cherry-pick -x 54fbd317f22ab072fe68488ff2375720465a844a~10..54fbd317f22ab072fe68488ff2375720465a844a --allow-empty --keep-redundant-commits`

A short term solution at the moment would be to make PRs with less than 22 commits as that is the number of commits in between the problem one. This is would only work until a change with files that only exist in upstream are pushed into midstream, causing a new problem.
The solution we should do though, would be changing the way magic-mirror copies our commits in general, instead of copying the number of commits inside of the PR it just copies the merged commit itself, as we are only pushing merge commits into main anyway.
Ex: `git cherry-pick -x fd29e047a161c0ff5f39fa9b14c2f9db70b4c0f0 --allow-empty --keep-redundant-commits`

xiangjingli commented:
/hold
